File: lib/pecos_rng/src/rng_pcg.c
/*
 * PCG Random Number Generation for C.
 *
 * Copyright 2014 Melissa O'Neill <oneill@pcg-random.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For additional information about the PCG random number generation scheme,
 * including its license and other licensing options, visit
 *
 *       http://www.pcg-random.org
 */

#include <math.h>
#include "rng_pcg.h"

// RNG state structure
typedef struct pcg_state_setseq_64 {
    uint64_t state;
    uint64_t inc;
} pcg32_random_t;

// global RNG state
static pcg32_random_t pcg32_global = {
    0x853c49e6748fea9bULL,
    0xda3e39cb94b95bdbULL
};

// default multi[plier]
#define PCG_DEFAULT_MULTIPLIER_64  6364136223846793005ULL

// helper functions
static inline uint32_t pcg_rotr_32(uint32_t value, unsigned int urot) {
    int rot = (int)urot;
    return (value >> rot) | (value << ((-rot) & 31));
}

static inline void pcg_setseq_64_step_r(pcg32_random_t* rng) {
    rng->state = rng->state * PCG_DEFAULT_MULTIPLIER_64 + rng->inc;
}

static inline uint32_t pcg_output_xsh_rr_64_32(uint64_t state) {
    return pcg_rotr_32(((state >> 18u) ^ state) >> 27u, state >> 59u);
}

static inline uint32_t pcg32_random_r(pcg32_random_t* rng) {
    const uint64_t oldstate = rng->state;
    pcg_setseq_64_step_r(rng);
    return pcg_output_xsh_rr_64_32(oldstate);
}

static inline uint32_t pcg32_boundedrand_r(pcg32_random_t* rng, uint32_t ubound) {
    int32_t bound = (int32_t)ubound;
    uint32_t threshold = -bound % bound;
    for (;;) {
        const uint32_t r = pcg32_random_r(rng);
        if (r >= threshold)
            return r % bound;
    }
}

static inline void pcg32_srandom_r(pcg32_random_t* rng, uint64_t initstate, uint64_t initseq) {
    rng->state = 0U;
    rng->inc = (initseq << 1u) | 1u;
    pcg_setseq_64_step_r(rng);
    rng->state += initstate;
    pcg_setseq_64_step_r(rng);
}

// public interface to RNG

uint32_t pcg32_random() {
    return pcg32_random_r(&pcg32_global);
}

uint32_t pcg32_boundedrand(uint32_t bound) {
    return pcg32_boundedrand_r(&pcg32_global, bound);
}

double pcg32_frandom() {
    return ldexp(pcg32_random(), -32);
}

void pcg32_srandom(uint64_t seq) {
    pcg32_srandom_r(&pcg32_global, 42u, seq);
}

File: crates/pecos-rng/src/rng_pcg.rs
#[derive(Clone, Copy)]
#[allow(clippy::cast_possible_truncation)]
#[allow(clippy::cast_possible_wrap)]
#[allow(clippy::cast_sign_loss)]
pub struct PCGRandom {
    pub state: u64,
    inc: u64,
}

impl PCGRandom {
    #[must_use]
    pub fn init_global_state() -> PCGRandom {
        PCGRandom {
            state: 0x853c_49e6_748f_ea9b,
            inc: 0xda3e_39cb_94b9_5bdb,
        }
    }

    #[inline]
    #[allow(clippy::cast_possible_wrap)]
    fn pcg_rotr(value: u32, urot: u32) -> u32 {
        let rot = urot as i32;
        (value >> rot) | (value << ((-rot) & 31))
    }

    #[inline]
    fn pcg_setseq_64_step_r(rng: &mut PCGRandom) {
        const PCG_DEFAULT_MULTIPLIER_64: u64 = 6_364_136_223_846_793_005;
        rng.state = rng
            .state
            .wrapping_mul(PCG_DEFAULT_MULTIPLIER_64)
            .wrapping_add(rng.inc);
    }

    #[inline]
    #[allow(clippy::cast_possible_truncation)]
    fn pcg_output_xsh(state: u64) -> u32 {
        let value = ((state >> 18) ^ state) >> 27;
        let urot = state >> 59;
        PCGRandom::pcg_rotr(value as u32, urot as u32)
    }

    #[inline]
    pub fn pcg32_random_r(rng: &mut PCGRandom) -> u32 {
        let old_state: u64 = rng.state;
        PCGRandom::pcg_setseq_64_step_r(rng);
        PCGRandom::pcg_output_xsh(old_state)
    }

    #[inline]
    #[allow(clippy::cast_possible_wrap, clippy::cast_sign_loss)]
    pub fn pcg32_boundedrand_r(rng: &mut PCGRandom, ubound: u32) -> u32 {
        let bound: i32 = ubound as i32;
        let threshold: u32 = (-bound % bound) as u32;
        loop {
            let random: u32 = PCGRandom::pcg32_random_r(rng);
            if random >= threshold {
                return random % bound as u32;
            }
        }
    }

    #[inline]
    pub fn frandom(rng: &mut PCGRandom) -> f64 {
        let random = f64::from(PCGRandom::pcg32_random_r(rng));
        let exp: i32 = -32;
        random * 2f64.powi(exp)
    }

    #[inline]
    pub fn pcg32_srandom_r(rng: &mut PCGRandom, initstate: u64, initseq: u64) {
        rng.state = 0_u64;
        rng.inc = (initseq << 1_u64) | 1_u64;
        PCGRandom::pcg_setseq_64_step_r(rng);
        rng.state += initstate;
        PCGRandom::pcg_setseq_64_step_r(rng);
    }
}
